File: sdk/ruby/kar98k.rb
# kar98k Ruby SDK — write load tests in native Ruby.
#
# Usage:
#   require_relative 'kar98k'
#
#   scenario name: "my-test", pattern: chaos(preset: "aggressive")
#
#   def setup
#     { token: "abc" }
#   end
#
#   def default(data)
#     resp = Http.get("http://localhost:8080/api")
#     check resp, "status 200" => ->(r) { r.status == 200 }
#   end

require 'json'

module Kar98k
  # --- Protocol layer ---

  def self._send(msg)
    $stdout.puts(JSON.generate(msg))
    $stdout.flush
  end

  def self._recv
    line = $stdin.gets
    exit(0) if line.nil?
    JSON.parse(line.strip)
  end

  # --- Response ---

  class Response
    attr_reader :status, :body, :duration, :error

    def initialize(status:, body: "", duration: 0.0, error: "")
      @status = status
      @body = body
      @duration = duration
      @error = error
    end

    def json
      JSON.parse(@body) rescue nil
    end

    def to_s
      "<Response status=#{@status} duration=#{@duration.round(3)}s>"
    end
  end

  # --- HTTP module ---

  module Http
    def self.get(url, **opts);    _request("GET", url, **opts); end
    def self.post(url, **opts);   _request("POST", url, **opts); end
    def self.put(url, **opts);    _request("PUT", url, **opts); end
    def self.delete(url, **opts); _request("DELETE", url, **opts); end
    def self.patch(url, **opts);  _request("PATCH", url, **opts); end

    def self._request(method, url, headers: nil, json: nil, body: nil)
      msg = { type: "http", method: method, url: url }
      if headers
        msg[:headers] = headers
      end
      if json
        msg[:body] = JSON.generate(json)
        msg[:headers] = (msg[:headers] || {}).merge("Content-Type" => "application/json")
      elsif body
        msg[:body] = body
      end

      Kar98k._send(msg)

      resp_data = Kar98k._recv
      Response.new(
        status: resp_data["status"] || 0,
        body: resp_data["body"] || "",
        duration: resp_data["duration"] || 0.0,
        error: resp_data["error"] || ""
      )
    end
  end

  # --- Config state ---

  @scenario_config = {}
  @chaos_config = {}
  @stages = []
  @thresholds = {}

  def self.scenario_config; @scenario_config; end
  def self.chaos_config; @chaos_config; end
  def self.stages_config; @stages; end
  def self.thresholds_config; @thresholds; end

  # --- Main loop ---

  def self.run_loop(binding_context)
    loop do
      cmd = _recv

      case cmd["cmd"]
      when "init"
        scenario_msg = { type: "scenario", name: @scenario_config[:name] || "" }
        scenario_msg[:chaos] = @chaos_config unless @chaos_config.empty?
        scenario_msg[:stages] = @stages unless @stages.empty?
        scenario_msg[:thresholds] = @thresholds unless @thresholds.empty?
        _send(scenario_msg)
        _send({ type: "done" })

      when "setup"
        data = if binding_context.respond_to?(:setup, true)
                 binding_context.send(:setup)
               else
                 {}
               end
        _send({ type: "done", data: data || {} })

      when "iterate"
        data = cmd["data"] || {}
        begin
          if binding_context.respond_to?(:default, true)
            binding_context.send(:default, data)
          elsif binding_context.respond_to?(:run, true)
            binding_context.send(:run, data)
          end
        rescue => e
          _send({ type: "error", message: e.message })
        end
        _send({ type: "done" })

      when "teardown"
        data = cmd["data"] || {}
        binding_context.send(:teardown, data) if binding_context.respond_to?(:teardown, true)
        _send({ type: "done" })

      else
        _send({ type: "error", message: "unknown command: #{cmd['cmd']}" })
      end
    end
  end
end

# --- Top-level DSL methods (available in user scripts) ---

def scenario(name:, pattern: nil, vus: nil, thresholds: nil)
  Kar98k.scenario_config[:name] = name
  Kar98k.instance_variable_set(:@chaos_config, pattern) if pattern
  Kar98k.instance_variable_set(:@stages, vus) if vus
  Kar98k.instance_variable_set(:@thresholds, thresholds) if thresholds
end

def chaos(preset: "moderate", spike_factor: nil, noise_amplitude: nil, lambda_: nil)
  cfg = { "preset" => preset }
  cfg["spike_factor"] = spike_factor if spike_factor
  cfg["noise_amplitude"] = noise_amplitude if noise_amplitude
  cfg["lambda"] = lambda_ if lambda_
  cfg
end

def stage(duration, target)
  { "duration" => duration, "target" => target }
end

def ramp(stages)
  stages
end

def check(response, checks)
  all_passed = true
  checks.each do |name, fn|
    passed = begin
               !!fn.call(response)
             rescue
               false
             end
    Kar98k._send({ type: "check", name: name, passed: passed })
    all_passed = false unless passed
  end
  all_passed
end

def sleep_dur(duration)
  if duration.is_a?(String)
    sleep(_parse_duration(duration))
  else
    sleep(duration)
  end
end

def think_time(min_dur, max_dur)
  min_s = _parse_duration(min_dur)
  max_s = _parse_duration(max_dur)
  rand(min_s..max_s)
end

def _parse_duration(s)
  return s.to_f if s.is_a?(Numeric)
  s = s.strip
  if s.end_with?("ms")
    s[0..-3].to_f / 1000
  elsif s.end_with?("us") || s.end_with?("µs")
    s[0..-3].to_f / 1_000_000
  elsif s.end_with?("m") && !s.end_with?("ms")
    s[0..-2].to_f * 60
  elsif s.end_with?("h")
    s[0..-2].to_f * 3600
  elsif s.end_with?("s")
    s[0..-2].to_f
  else
    s.to_f
  end
end

# Alias Http to top-level
Http = Kar98k::Http

# Auto-start protocol loop when script finishes loading
at_exit do
  next if $stdin.tty?
  Kar98k.run_loop(self)
end

File: sdk/python/kar98k.py
"""kar98k Python SDK — write load tests in native Python.

Usage:
    from kar98k import scenario, chaos, http, check, sleep, think_time, stage, ramp

    scenario(name="my-test", pattern=chaos(preset="aggressive"))

    def setup():
        return {"token": "abc"}

    def default(data):
        resp = http.get("http://localhost:8080/api", headers={"Authorization": f"Bearer {data['token']}"})
        check(resp, {"status 200": lambda r: r.status == 200})
        sleep(think_time("1s", "3s"))
"""

import sys
import json
import random
import time as _time
import inspect
import atexit

# --- Protocol layer (hidden from user) ---

def _send(msg):
    print(json.dumps(msg), flush=True)

def _recv():
    line = sys.stdin.readline()
    if not line:
        sys.exit(0)
    return json.loads(line.strip())


# --- Response object ---

class Response:
    """HTTP response returned by http.get/post/etc."""

    def __init__(self, status, body="", duration=0.0, error=""):
        self.status = status
        self.body = body
        self.duration = duration
        self.error = error

    def json(self):
        return json.loads(self.body) if self.body else None

    def __repr__(self):
        return f"<Response status={self.status} duration={self.duration:.3f}s>"


# --- HTTP module ---

class _HTTP:
    """HTTP client — http.get(), http.post(), etc."""

    def get(self, url, **kwargs):
        return self._request("GET", url, **kwargs)

    def post(self, url, **kwargs):
        return self._request("POST", url, **kwargs)

    def put(self, url, **kwargs):
        return self._request("PUT", url, **kwargs)

    def delete(self, url, **kwargs):
        return self._request("DELETE", url, **kwargs)

    def patch(self, url, **kwargs):
        return self._request("PATCH", url, **kwargs)

    def _request(self, method, url, headers=None, json_body=None, body=None):
        msg = {"type": "http", "method": method, "url": url}
        if headers:
            msg["headers"] = headers
        if json_body is not None:
            msg["body"] = json.dumps(json_body)
            if not headers:
                msg["headers"] = {}
            msg["headers"]["Content-Type"] = "application/json"
        elif body:
            msg["body"] = body

        _send(msg)

        # Read response from kar98k
        resp_data = _recv()
        return Response(
            status=resp_data.get("status", 0),
            body=resp_data.get("body", ""),
            duration=resp_data.get("duration", 0.0),
            error=resp_data.get("error", ""),
        )


http = _HTTP()


# --- Scenario config ---

_scenario_config = {}
_chaos_config = {}
_stages = []
_thresholds = {}


def scenario(name, pattern=None, vus=None, thresholds=None):
    """Declare test scenario configuration."""
    global _scenario_config, _chaos_config, _stages, _thresholds
    _scenario_config = {"name": name}
    if pattern:
        _chaos_config = pattern
    if vus:
        _stages = vus
    if thresholds:
        _thresholds = thresholds


def chaos(preset="moderate", spike_factor=None, noise_amplitude=None, lambda_=None):
    """Configure chaos traffic patterns."""
    cfg = {"preset": preset}
    if spike_factor is not None:
        cfg["spike_factor"] = spike_factor
    if noise_amplitude is not None:
        cfg["noise_amplitude"] = noise_amplitude
    if lambda_ is not None:
        cfg["lambda"] = lambda_
    return cfg


def stage(duration, target):
    """Define a VU ramping stage."""
    return {"duration": duration, "target": target}


def ramp(stages):
    """Wrap stages for VU ramping."""
    return stages


# --- Check ---

def check(response, checks):
    """Run assertions against a response.

    Args:
        response: Response object from http.*
        checks: dict of {"name": lambda r: bool}
    """
    all_passed = True
    for name, fn in checks.items():
        try:
            passed = bool(fn(response))
        except Exception:
            passed = False
        _send({"type": "check", "name": name, "passed": passed})
        if not passed:
            all_passed = False
    return all_passed


# --- Sleep / Think Time ---

def sleep(duration):
    """Sleep for a duration string (e.g., '1s', '500ms') or seconds (float)."""
    if isinstance(duration, str):
        _time.sleep(_parse_duration(duration))
    else:
        _time.sleep(float(duration))


def think_time(min_dur, max_dur):
    """Generate a random duration between min and max (chaos-aware)."""
    min_s = _parse_duration(min_dur)
    max_s = _parse_duration(max_dur)
    return random.uniform(min_s, max_s)


def _parse_duration(s):
    """Parse Go-style duration string to seconds."""
    if isinstance(s, (int, float)):
        return float(s)
    s = s.strip()
    if s.endswith("ms"):
        return float(s[:-2]) / 1000
    elif s.endswith("us") or s.endswith("µs"):
        return float(s[:-2]) / 1_000_000
    elif s.endswith("s") and not s.endswith("ms"):
        return float(s[:-1])
    elif s.endswith("m"):
        return float(s[:-1]) * 60
    elif s.endswith("h"):
        return float(s[:-1]) * 3600
    return float(s)


# --- Group ---

def group(name, fn):
    """Group related requests for metric labeling."""
    return fn()


# --- Main loop (auto-starts when script is run by kar98k) ---

_caller_module = None

# Capture the importing module's globals at import time
for _frame_info in inspect.stack():
    _f = _frame_info[0]
    if _f.f_globals.get("__name__") == "__main__":
        _caller_module = _f.f_globals
        break


def _main():
    global _caller_module
    if _caller_module is None:
        _caller_module = sys.modules.get("__main__").__dict__ if "__main__" in sys.modules else {}

    setup_fn = _caller_module.get("setup")
    default_fn = _caller_module.get("default") or _caller_module.get("run")
    teardown_fn = _caller_module.get("teardown")

    if default_fn is None:
        _send({"type": "error", "message": "script must define a default() or run() function"})
        sys.exit(1)

    while True:
        cmd = _recv()

        if cmd["cmd"] == "init":
            scenario_msg = {"type": "scenario", "name": _scenario_config.get("name", "")}
            if _chaos_config:
                scenario_msg["chaos"] = _chaos_config
            if _stages:
                scenario_msg["stages"] = _stages
            if _thresholds:
                scenario_msg["thresholds"] = _thresholds
            _send(scenario_msg)
            _send({"type": "done"})

        elif cmd["cmd"] == "setup":
            data = None
            if setup_fn:
                data = setup_fn()
            _send({"type": "done", "data": data or {}})

        elif cmd["cmd"] == "iterate":
            data = cmd.get("data", {})
            try:
                default_fn(data)
            except Exception as e:
                _send({"type": "error", "message": str(e)})
            _send({"type": "done"})

        elif cmd["cmd"] == "teardown":
            data = cmd.get("data", {})
            if teardown_fn:
                try:
                    teardown_fn(data)
                except Exception:
                    pass
            _send({"type": "done"})

        else:
            _send({"type": "error", "message": f"unknown command: {cmd['cmd']}"})


# Auto-start: when imported by a kar98k subprocess, block on the main loop
# after the importing module finishes defining setup/default/teardown.
def _auto_start():
    """Called via atexit — runs the protocol loop on the main thread."""
    # Only activate when stdin is a pipe (i.e., run by kar98k)
    if sys.stdin.isatty():
        return
    _main()

atexit.register(_auto_start)
